File: 1843 (881 Div 3)/D_apple.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
 
vector<vector<int>> adj;
vector<ll> cnt;

void dfs(int curr, int par){
  if (adj[curr].size() == 1 && adj[curr][0] == par){
    cnt[curr] = 1;
  }
  else{
    for (auto nei: adj[curr]){
      if(nei != par){
        dfs(nei, curr);
        cnt[curr] += cnt[nei];
      }
    }
  }
}


void solve(){
  int n, q;
  cin >> n;

  adj.assign(n,vector<int>());

  for (int i = 0; i < n-1; i++){
    int u, v;

    cin >> u >> v;
    u--;
    v--;

    adj[u].push_back(v);
    adj[v].push_back(u);

  }

  cnt.assign(n, 0);

  dfs(0, -1);

  cin >> q;
  for (int i = 0; i<q;i++){
    int a,b;

    cin >> a >> b;
    a--;
    b--;

    ll res = cnt[a] * cnt[b];

    cout << res << '\n';
  }
}

int main(){
  int tc;

  cin >> tc;

  while (tc--){
    solve();
  }
  return 0;
}

File: 1843 (881 Div 3)/D_apple.py
import sys
input = sys.stdin.readline
from collections import defaultdict
def solve():
    n = int(input())
    adj = defaultdict(list)
    for edge in range(n-1):
        a,b = map(int, input().strip().split())
        adj[a].append(b)
        adj[b].append(a)
    
    cnt = [0] * (n+1)
    
    def dfs(curr, par):
        
        if len(adj[curr]) == 1 and adj[curr][0] == par:
            cnt[curr] = 1
            return
        for nei in adj[curr]:
            dfs(nei, curr)
            cnt[curr] += cnt[nei]
    dfs(1, -1)
    for q in range(int(input())):
        a,b = map(int, input().strip().split())

        print(cnt[a] * cnt[b])

for tc in range(int(input())):
    solve()